File: bindings/java/src/main/java/org/indunet/robot/bus/Publisher.java
package org.indunet.robot.bus;

import com.sun.jna.Pointer;

/** Standalone pub socket (not tied to a {@link Node}). */
public final class Publisher implements AutoCloseable {
    private Pointer ptr;

    public Publisher() {
        this((String) null);
    }

    public Publisher(String endpoint) {
        this.ptr = Errors.checkPtr(RobotBusC.Holder.INSTANCE.robot_bus_publisher_new(endpoint), "Publisher");
    }

    public void publish(String topic, byte[] payload) {
        Errors.check(
                RobotBusC.Holder.INSTANCE.robot_bus_publisher_publish(
                        ptr, topic, payload, payload.length),
                "publish");
    }

    public String endpoint() {
        return NativeUtils.takeCString(RobotBusC.Holder.INSTANCE.robot_bus_publisher_endpoint(ptr));
    }

    @Override
    public void close() {
        RobotBusC.Holder.INSTANCE.robot_bus_publisher_free(ptr);
        ptr = null;
    }
}

File: bindings/android/src/main/kotlin/org/indunet/robot/bus/Publisher.kt
package org.indunet.robot.bus

import com.sun.jna.Pointer

/** Standalone pub socket (not tied to a [Node]). */
class Publisher
@JvmOverloads
constructor(endpoint: String? = null) : AutoCloseable {
    private var ptr: Pointer? =
        Errors.checkPtr(RobotBusC.Holder.INSTANCE.robot_bus_publisher_new(endpoint), "Publisher")

    fun publish(topic: String, payload: ByteArray) {
        Errors.check(
            RobotBusC.Holder.INSTANCE.robot_bus_publisher_publish(
                ptr,
                topic,
                payload,
                payload.size.toLong(),
            ),
            "publish",
        )
    }

    fun endpoint(): String =
        NativeUtils.takeCString(RobotBusC.Holder.INSTANCE.robot_bus_publisher_endpoint(ptr))

    override fun close() {
        RobotBusC.Holder.INSTANCE.robot_bus_publisher_free(ptr)
        ptr = null
    }
}
